src/internal/connector/graph/service.go

```diff
@@ -319,3 +365,181 @@ func RunWithRetry(run func() error) error {
 
 	return support.ConnectorStackErrorTraceWrap(err, "maximum retries or unretryable")
 }
+
+// RetryHandler handles transient HTTP responses and retries the request given the retry options
+type RetryHandler struct {
+	// default options to use when evaluating the response
+	options RetryHandlerOptions
+}
+
+type RetryHandlerOptions struct {
+	// Callback to determine if the request should be retried
+	ShouldRetry func(delay time.Duration, executionCount int, request *http.Request, response *http.Response) bool
+	// The maximum number of times a request can be retried
+	MaxRetries int
+	// The delay in seconds between retries
+	DelaySeconds int
+}
+
+type retryHandlerOptionsInt interface {
+	abs.RequestOption
+	GetShouldRetry() func(delay time.Duration, executionCount int,
+		request *http.Request, response *http.Response) bool
+	GetDelaySeconds() int
+	GetMaxRetries() int
+}
+
+var retryKeyValue = abs.RequestOptionKey{
+	Key: "RetryHandler",
+}
+
+// GetKey returns the key value to be used when the option is added to the request context
+func (options *RetryHandlerOptions) GetKey() abs.RequestOptionKey {
+	return retryKeyValue
+}
+
+// GetShouldRetry returns the should retry callback function which evaluates the response for retrial
+func (options *RetryHandlerOptions) GetShouldRetry() func(delay time.Duration,
+	executionCount int, request *http.Request, response *http.Response) bool {
+	return options.ShouldRetry
+}
+
+// GetDelaySeconds returns the delays in seconds between retries
+func (options *RetryHandlerOptions) GetDelaySeconds() int {
+	if options.DelaySeconds < 1 {
+		return defaultDelaySeconds
+	}
+
+	return options.DelaySeconds
+}
+
+// GetMaxRetries returns the maximum number of times a request can be retried
+func (options *RetryHandlerOptions) GetMaxRetries() int {
+	if options.MaxRetries < 1 {
+		return defaultMaxRetries
+	}
+
+	return options.MaxRetries
+}
+
+// Intercept implements the interface and evaluates whether to retry a failed request.
+func (middleware RetryHandler) Intercept(
+	pipeline khttp.Pipeline,
+	middlewareIndex int,
+	req *http.Request,
+) (*http.Response, error) {
+	obsOptions := khttp.GetObservabilityOptionsFromRequest(req)
+	ctx := req.Context()
+
+	var (
+		span              trace.Span
+		observabilityName string
+	)
+
+	if obsOptions != nil {
```

Comment:
is telemetry required for the baseline implementation of the retry handler?  If not, we should save its addition for a follow-up PR.  Keep this one as small as possible.  Besides, we already have a telemetry package, and may want to integrate with that.